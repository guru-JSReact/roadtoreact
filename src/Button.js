import React from "react";
import PropTypes from "prop-types";

const Button = ({onclick, children}) => {
    return <button
        type="button"
        onClick={onclick}
    >{children}
    </button>
};

Button.prototype = {
    onClick: PropTypes.func,
    children: PropTypes.node,
}

export default Button;
