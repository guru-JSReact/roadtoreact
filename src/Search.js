import React from "react";

const Search = ({input, onChange, children}) =>
    <form>
        {children}
        <input
            type="text"
            value={input}
            onChange={onChange}
        />
    </form>;
export default Search;
