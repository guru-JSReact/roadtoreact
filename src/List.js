import React from "react";
import Button from "./Button";

const List = ({list, onDismiss}) => {
    return (
        <div>
            {
                list.map(item =>
                    <div key={item.objectID}>
                       <span>{item.title}</span>
                       <span>{item.author}</span>
                        <span>{item.num_comments}</span>
                        <span>{item.points}</span>
                        <span>
                            <Button onclick={()=>onDismiss(item.objectID)}>Dismiss </Button>
                        </span>
                    </div>

            )
            }
        </div>
    )
}
export default List;
