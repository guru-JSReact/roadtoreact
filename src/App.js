import React, {Component} from 'react';
import './App.css';
import Search from "./Search";
import Button from "./Button";
import List from "./List";


const list = [
    {
        title: 'React',
        url: 'https://facebook.github.io/react/',
        author: 'Jordan Walker',
        num_comments: 3,
        points: 4,
        objectID: 0,
    },
    {
        title: 'React1',
        url: 'https://facebook.github.io/react/',
        author: 'Jordan Walker1',
        num_comments: 31,
        points: 41,
        objectID: 1,
    },
    {
        title: 'React2',
        url: 'https://facebook.github.io/react/',
        author: 'Jordan Walker3',
        num_comments: 32,
        points: 42,
        objectID: 2,
    }
]

const isSearched = searchTerm => item => {
    console.log(item.title.toLowerCase().includes(searchTerm.toLowerCase()));
    !searchTerm || item.title.toLowerCase().includes(searchTerm.toLowerCase());
}

const DEFAULT_QUERY = 'redux';
const DEFAULT_HPP = '100';
const PATH_BASE = 'https://hn.algolia.com/api/v1';
const PATH_SEARCH = '/search';
const PARAM_SEARCH = 'query=';
const PARAM_PAGE = 'page=';
const PARAM_HPP = 'hitsPerPage=';

// const url = `${PATH_BASE}${PATH_SEARCH}?${PARAM_SEARCH}${DEFAULT_QUERY}`;


class App extends Component {
    constructor(props) {
        super(props);
        this.state = {
            result: '',
            searchKey: '',
            searchTerm: DEFAULT_QUERY,
        }
    }

    setSearchTopStories = (result) => {
        const {hits, page} = result;
        const {searchKey, results} = this.state;
        const oldHits = results && results[searchKey] ? results[searchKey].hits : [];
        const updatedHits = [...oldHits, ...hits];
        this.setState({
            result,
            [searchKey]: {hit: updatedHits, page}
        });

    }

    componentDidMount() {
        const {searchTerm} = this.state;
        this.setState({searchKey: searchTerm});
        this.fetchSearchTopStories(searchTerm);
    }
    needsToSearchTopStories = (searchTerm)  => !this.state.results[searchTerm];


    onDismiss = (id) => {
        const { searchKey, results } = this.state; const { hits, page } = results[searchKey];
        const isNotId = item => item.objectID !== id;
        const updatedHits = hits.filter(isNotId);
        this.setState({ results: {...results,
                [searchKey]: { hits: updatedHits, page }
            } });
    }

    fetchSearchTopStories = (searchTerm, page = 1) => {
        fetch(`${PATH_BASE}${PATH_SEARCH}?${PARAM_SEARCH}${searchTerm}&${PARAM_PAGE}${page}&${PARAM_HPP}${DEFAULT_HPP}`)
            .then(response => response.json())
            .then(result => this.setSearchTopStories(result))
            .catch(error => error);
    }

    onSearchHandler = (event) => {
        this.setState({searchTerm: event.target.value})
    }
    onSearchSubmit = (event) => {
        const {searchTerm} = this.state;
        this.setState({searchKey: searchTerm});
        this.fetchSearchTopStories(searchTerm);
        event.preventDefault();
        if (this.needsToSearchTopStories(searchTerm)) {
            this.fetchSearchTopStories(searchTerm);
        }
    }

    render() {
        const {result, searchTerm, searchKey} = this.state;
        const page = (result && result[searchKey] && result[searchKey].page) || 0;
        const list = (result && result[searchKey] && result[searchKey].hits) || [];
        return (
            <div className="page">
                <Search
                    value={searchTerm}
                    onChange={this.onSearchHandler}
                    onSubmit={this.onSearchSubmit}
                >Search</Search>
                {
                    result &&
                    <List list={list}
                          onDismiss={this.onDismiss}
                    />
                }
                <div className="interactions">
                    <Button onClick={() => this.fetchSearchTopStories(searchKey, page + 1)}>
                        More
                    </Button>
                </div>
            </div>
        );
    }
}

export default App;
