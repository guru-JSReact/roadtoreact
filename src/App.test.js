import React from 'react';
import ReactDOM from 'react-dom';
import {render} from '@testing-library/react';
import renderer from 'react-test-renderer';
import Search from "./Search";
import Button from "./Button";
import List from "./List";
import App from './App';
//
// test('page', () => {
//     const {getByText} = render(<App/>);
//     const linkElement = getByText(/page/);
//     expect(linkElement).toBeInTheDocument();
// });

describe('App', () => {
    it('renders without crashing', () => {
        const div = document.createElement('div');
        ReactDOM.render(<App/>, div);
        ReactDOM.unmountComponentAtNode(div);
    });

    test('has a valid snapshot', () => {
        const component = renderer.create(
            <App/>);
        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });
});

describe('Search', () => {
    it('renders without crashing', () => {
        const div = document.createElement('div'); ReactDOM.render(<Search>Search</Search>, div); ReactDOM.unmountComponentAtNode(div);
    });
    test('has a valid snapshot', () => { const component = renderer.create(
        <Search>Search</Search>
    );
        let tree = component.toJSON();
        expect(tree).toMatchSnapshot(); });
});

describe('Button', () => {
    it('renders without crashing', () => {
        const div = document.createElement('div'); ReactDOM.render(<Button>Give Me More</Button>, div); ReactDOM.unmountComponentAtNode(div);
    });
    test('has a valid snapshot', () => { const component = renderer.create(
        <Button>Give Me More</Button>
    );
        let tree = component.toJSON();
        expect(tree).toMatchSnapshot(); });
});

describe('List', () => {
    const props = {
        list: [
            { title: '1', author: '1', num_comments: 1, points: 2, objectID: 'y' },
            { title: '2', author: '2', num_comments: 1, points: 2, objectID: 'z' }, ],
    };

    it('renders without crashing', () => {
        const div = document.createElement('div'); ReactDOM.render(<Table { ...props } />, div);
    });
    test('has a valid snapshot', () => { const component = renderer.create(
        <List { ...props } /> );
        let tree = component.toJSON();
        expect(tree).toMatchSnapshot(); });
});


