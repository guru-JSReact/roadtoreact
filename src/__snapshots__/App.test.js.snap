// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App has a valid snapshot 1`] = `
<div
  className="page"
>
  <form
    onSubmit={[Function]}
  >
    <input
      onChange={[Function]}
      type="text"
      value="redux"
    />
    <button
      type="submit"
    >
      Search
    </button>
  </form>
  
  <div
    className="interactions"
  >
    <button
      type="button"
    >
      More
    </button>
  </div>
</div>
`;

exports[`Button has a valid snapshot 1`] = `
<button
  type="button"
>
  Give Me More
</button>
`;

exports[`List has a valid snapshot 1`] = `
<div>
  <div>
    <span>
      1
    </span>
    <span>
      1
    </span>
    <span>
      1
    </span>
    <span>
      2
    </span>
    <span>
      <button
        onClick={[Function]}
        type="button"
      >
        Dismiss 
      </button>
    </span>
  </div>
  <div>
    <span>
      2
    </span>
    <span>
      2
    </span>
    <span>
      1
    </span>
    <span>
      2
    </span>
    <span>
      <button
        onClick={[Function]}
        type="button"
      >
        Dismiss 
      </button>
    </span>
  </div>
</div>
`;

exports[`Search has a valid snapshot 1`] = `
<form>
  <input
    type="text"
  />
  <button
    type="submit"
  >
    Search
  </button>
</form>
`;
